### PR TITLE
Remote ota/development

### PR DESCRIPTION
**Pull request by chezlyne**
Adjusted the Logic for write into the Flash memory of the Robot.

### CODE REVIEW

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 1]
```diff
@@ -94,29 +99,251 @@ class IntProgCmdProvider: public CmdProvider {
...
+     * @param rsp Reference to a pointer to the next response.
+     * @return True if there is a next command and next response, false otherwise.
+     */
+    virtual bool next(const CommandInfo*& cmd, const ResponseInfo*& rsp)
+    {
+        /*Keep original command.*/
+        const CommandInfo* originalCmd = &m_cmds[nextCmd];
+
+        /*Copy original command.*/
+        CommandInfo updatedCmd = *originalCmd;
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:165]: not needed as having no value add. Use m_updateCmd directly below.

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 2]
```diff
@@ -94,29 +99,251 @@ class IntProgCmdProvider: public CmdProvider {
...
+
+        /*Update the command with required content to write in the flash memory.*/
+        updateFwProgCmd(&updatedCmd);
+
+        /*Save the updated command information.*/
+        m_updatedCmd = updatedCmd;
+
+        cmd = &m_updatedCmd;
+
+        if(nullptr != cmd)
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:175]: explain me how this can be null ?

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 3]
```diff
@@ -94,29 +99,251 @@ class IntProgCmdProvider: public CmdProvider {
...
+     * This variable represents the current memory address where data will be written
+     * during the firmware programming process. It is used to keep track of the
+     * memory address for sequential data writes.
+     */
+    uint16_t  m_currWriteMemAddr; 
+
+   /**
+    * @brief Updates the command according to specific requirements.
+    *
+    * This method updates the provided command based on specific requirements
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:246]: I don't see any requirements it is basefd on.  Just comment what it does and why.

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 4]
```diff
@@ -94,29 +99,251 @@ class IntProgCmdProvider: public CmdProvider {
...
+    */
+    void updateFwProgCmd(CommandInfo* fwProgCmd)
+    {
+        if(nullptr != fwProgCmd->command)
+        {   
+            if (fwProgCmd->command == Zumo32U4Specification::SET_MEMORY_ADDR)
+            {
+                /**<Create a buffer for the updated information.*/
+                m_updatedCmdBufferSize = 3;/**< Final Size of the updated command for SET_MEMORY_ADR.*/
+                delete[] m_updatedCmdBuffer; /**<Free existing buffer.*/
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:263]: This new/delete is poison for embedded system code. Just use a buffer array that is large enough for the longest command in m_updatedCmdBuffer 128 + x-

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 5]
```diff
@@ -227,13 +468,45 @@ bool BootloaderCom::process()
             m_waitingForResponse = false;
             if(m_cmdProvider->next(m_currentCommand, m_currentResponse))
             {
-                if(true == m_flashManager.sendCommand(
-                    m_currentCommand->command,
-                    m_currentCommand->commandsize))
+                /*Check if the command provider's command type is for firmware programming.*/
+                if (FW_PROG == m_cmdProvider->getCmdType())
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:472]: This if doesn not belong here. This function is just handlgng cmd data exchanging. It must not know what cmds or handle them conditionally.

File: lib/ZumoOta/src/BootloaderCom.h [hunk 6]
```diff
@@ -193,13 +222,14 @@ class BootloaderCom
     /** 
      *Specifies the current memory/page address for flashing the firmware of the Zumo program flash.
      */
-    uint16_t m_currWriteMemAddr;
+    uint16_t static m_currWriteMemAddr;
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.h:225]: why static ??

File: lib/ZumoOta/src/FileManager.cpp [hunk 7]
```diff
@@ -80,3 +80,38 @@ bool FileManager::init()
     }
     return true;
 }
+
+size_t FileManager::readBytes(const char* firmwareName, uint8_t* buffer)
```
Comment (nhjschulz) [inline lib/ZumoOta/src/FileManager.cpp:84]: This is broken in several ways.

 * a read function does't open /close files, it operates on an open file handle
 * you need to pass the number of bytes to read, and return what you really got
 * Read function doesn't seek, it just keeps reading the file handle.

How do you detect when you read all bytes  of the bytes.

File: lib/ZumoOta/src/BootloaderCom.h [hunk 8]
```diff
@@ -99,7 +113,22 @@ class CmdProvider {
      *@param rsp Reference to a pointer to the next response.
      *@return True if there is a next command and response pair, otherwise false.
      */
-    virtual bool next(const CommandInfo *& cmd,  const ResponseInfo *& rsp) = 0;
+    virtual bool next( const CommandInfo *& cmd, const ResponseInfo *& rsp) = 0;
+
+    /**
+     *@brief Read 128 bytes from the file system.
+     */
+    static uint8_t* m_buffer;
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.h:121]: why a pointer ?

**Review by nhjschulz — CHANGES_REQUESTED** [review]
several issues

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 1 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp]: It will be take care by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 3 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:256]: It will be take care by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 4 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp]: It will be take care by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 5 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp]: It will be take care by the next commit.

File: lib/ZumoOta/src/BootloaderCom.h (hunk 8 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.h:109]: It will be take care by the next commit.

File: lib/ZumoOta/src/FileManager.cpp (hunk 7 above)
Comment (chezlyne) [reply lib/ZumoOta/src/FileManager.cpp]: It will be adjust  by the next commit.

**Comment by BlueAndi** [comment]
@chezlyne Why did you close it and opened a new one? Please stay with the PR otherwise all comments are lost and its hard for the reviewer to follow.

**Pull request closed without merging**